src/presentation/service/services.go

```diff
@@ -0,0 +1,449 @@
+		}
+		startupFilePtr = &startupFile
+	}
+
+	envs := []valueObject.ServiceEnv{}
+	if input["envs"] != nil {
+		rawEnvs, assertOk := input["envs"].([]string)
+		if !assertOk {
+			return NewServiceOutput(UserError, "EnvsMustBeStringArray")
+		}
+
+		for _, rawEnv := range rawEnvs {
+			env, err := valueObject.NewServiceEnv(rawEnv)
+			if err != nil {
+				slog.Debug(err.Error(), slog.String("env", rawEnv))
+				continue
+			}
+			envs = append(envs, env)
+		}
+	}
+
+	portBindings := []valueObject.PortBinding{}
+	if input["portBindings"] != nil {
+		rawPortBindings, assertOk := input["portBindings"].([]string)
+		if !assertOk {
+			return NewServiceOutput(UserError, "PortBindingsMustBeStringArray")
+		}
+
+		for _, rawPortBinding := range rawPortBindings {
+			portBinding, err := valueObject.NewPortBinding(rawPortBinding)
+			if err != nil {
+				slog.Debug(err.Error(), slog.String("portBinding", rawPortBinding))
+				continue
+			}
+			portBindings = append(portBindings, portBinding)
+		}
+	}
+
+	var autoStartPtr *bool
+	if input["autoStart"] != nil {
+		autoStart, err := voHelper.InterfaceToBool(input["autoStart"])
+		if err != nil {
+			return NewServiceOutput(UserError, "AutoStartMustBeBool")
+		}
+		autoStartPtr = &autoStart
+	}
+
+	var timeoutStartSecsPtr *uint
+	if input["timeoutStartSecs"] != nil {
+		timeoutStartSecs, err := voHelper.InterfaceToUint(input["timeoutStartSecs"])
+		if err != nil {
+			return NewServiceOutput(UserError, "TimeoutStartSecsMustBeUint")
+		}
+		timeoutStartSecsPtr = &timeoutStartSecs
+	}
+
+	var autoRestartPtr *bool
+	if input["autoRestart"] != nil {
+		autoRestart, err := voHelper.InterfaceToBool(
+			input["autoRestart"],
+		)
+		if err != nil {
+			return NewServiceOutput(UserError, "AutoRestartMustBeBool")
+		}
+		autoRestartPtr = &autoRestart
+	}
+
+	var maxStartRetriesPtr *uint
+	if input["maxStartRetries"] != nil {
+		maxStartRetries, err := voHelper.InterfaceToUint(input["maxStartRetries"])
+		if err != nil {
+			return NewServiceOutput(UserError, "MaxStartRetriesMustBeUint")
+		}
+		maxStartRetriesPtr = &maxStartRetries
+	}
+
+	autoCreateMapping := true
+	if input["autoCreateMapping"] != nil {
+		autoCreateMapping, err = voHelper.InterfaceToBool(input["autoCreateMapping"])
+		if err != nil {
+			return NewServiceOutput(UserError, "AutoCreateMappingMustBeBool")
+		}
+	}
+
+	if shouldSchedule {
+		cliCmd := infraEnvs.SpeediaOsBinary + " services create-installable"
+		installParams := []string{
+			"--name", name.String(),
+			"--auto-create-mapping", strconv.FormatBool(autoCreateMapping),
+		}
+
+		if len(envs) > 0 {
+			for _, env := range envs {
+				escapedField := shellescape.Quote(env.String())
+				installParams = append(installParams, "--envs", escapedField)
+			}
+		}
+
+		if len(portBindings) > 0 {
+			for _, portBinding := range portBindings {
+				escapedField := shellescape.Quote(portBinding.String())
+				installParams = append(installParams, "--port-bindings", escapedField)
+			}
+		}
```

Comment:
I may be wrong but those len checks are useless as if there are nothing on the range variable, the loop just won't run.

Reply:
You are correct.
However, checking the size before the loop is not to avoid errors but to improve code understanding. Note that all props that are optional are checked before they are actually added to the **CLI** command. The idea here is to maintain linearity in understanding, so it will not harm execution or considerably increase the code size, resulting in almost zero complexity and providing benefits for readability.